lib/spack/spack/environment/environment.py

```diff
@@ -1794,31 +1797,73 @@ def _add_concrete_spec(self, spec, concrete, new=True):
+          the in-development spec y, and then later reinstall the whole
+          environment.
+        """
+        db = _database or spack.store.STORE.db
 
-        # Collect their hashes, and the hashes of their installed parents.
-        # Notice: with order=breadth all changed dev specs are at depth 0,
-        # even if they occur as parents of one another.
-        return [
-            spec.dag_hash()
-            for depth, spec in traverse.traverse_nodes(
-                changed_dev_specs,
-                root=True,
-                order="breadth",
-                depth=True,
-                direction="parents",
-                key=traverse.by_dag_hash,
-            )
-            if depth == 0 or spec.installed
-        ]
+        overwrite_specs = set()
+        transitive_dev_install_times = collections.defaultdict(int)
+
+        def update_transitive_dev_install_times(spec):
+            dev_path_var = spec.variants.get("dev_path", None)
+            if dev_path_var and db.installed(spec):
+                _, record = db.query_by_spec_hash(spec.dag_hash())
+                when_this_spec_was_installed = record.installation_time
+
+                transitive_dev_install_times[spec] = max(
+                    transitive_dev_install_times[spec], when_this_spec_was_installed
+                )
+            # else: this is not a develop spec, and we don't want to record the
+            # installation time (we could have done an overwrite install of a
+            # non-develop spec, in which case it should not be necessary to
+            # reinstall the parent)
+            transitive_dev_install_time_child = transitive_dev_install_times[spec]
+            for parent in spec.dependents():
+                transitive_dev_install_times[parent] = max(
+                    transitive_dev_install_times[parent], transitive_dev_install_time_child
+                )
+
+        for spec in traverse.traverse_nodes(
+            self.concrete_roots(), direction="children", order="post"
+        ):
+            # If it is a parent of an overwite spec (transitive), then
+            # overwrite
+            if any(child in overwrite_specs for child in spec.dependencies()):
+                overwrite_specs.add(spec)
+                continue
+
+            update_transitive_dev_install_times(spec)
+
+            # Say x->y->z, you force uninstall y, then you change the source
+            # of z (which is develop). x should be reinstalled, but not because
+            # of y. Therefore, we wait until now to bail on uninstalled specs.
+            if not db.installed(spec):
+                continue
+
+            # If a transitive child develop spec has a greater install
+            # time, then overwrite
+            _, record = db.query_by_spec_hash(spec.dag_hash())
+            when_this_spec_was_installed = record.installation_time
+            if when_this_spec_was_installed < transitive_dev_install_times[spec]:
```

Comment:
Instead of looking up this value yet again, why not return it from `update_transitive_dev_install_times`